contracts/os/andromeda-kernel/src/execute.rs

```diff
@@ -918,251 +918,260 @@ pub fn unset_env(execute_ctx: ExecuteContext, variable: String) -> Result<Respon
+        username,
+        address: origin_address,
+        from_chain: current_chain,
+        to_chain: destination_chain.clone(),
+        funds: vec![], // Assuming funds are not handled here as per the given instructions
+        channel: channel.to_string(),
+    }
+}
+
+pub(crate) fn handle_ibc_direct(
+    deps: DepsMut,
+    info: MessageInfo,
+    env: Env,
+    ctx: Option<AMPPkt>,
+    message: AMPMsg,
+    channel_info: ChannelInfo,
+) -> Result<Response, ContractError> {
+    println!("Starting handle_ibc_direct function");
+
+    // Validate that the message is not empty
+    if message.message.is_empty() {
+        println!("Message is empty, returning error");
+        return Err(ContractError::InvalidPacket {
+            error: Some("Cannot send an empty message without funds via IBC".to_string()),
+        });
+    }
+
+    println!("Message is not empty, proceeding");
+
+    // Extract the destination chain from the recipient
+    let destination_chain = message
+        .recipient
+        .get_chain()
+        .expect("Destination chain not found in recipient");
+    println!("Extracted destination chain: {}", destination_chain);
+
+    // Retrieve the direct channel ID for the destination chain
+    let channel = channel_info.direct_channel_id.expect(&format!(
+        "Channel not found for chain {}",
+        &destination_chain
+    ));
+
+    // Build IBC context
+    let current_chain = CURR_CHAIN.load(deps.storage)?;
+    println!("Loaded current chain: {}", current_chain);
+
+    let amp_ctx = build_ibc_context(
+        &deps,
+        &info,
+        &env,
+        &channel,
+        ctx,
+        current_chain,
+        message.clone(),
+        &destination_chain.to_string(),
+    )?;
+    println!("Built IBC context for direct send");
+
+    // Create and send IBC message
+    let kernel_msg = IbcExecuteMsg::SendMessage {
+        amp_packet: amp_ctx,
+    };
+    println!("Created kernel message for sending");
+
+    let msg = IbcMsg::SendPacket {
+        channel_id: channel.clone(),
+        data: to_json_binary(&kernel_msg)?,
+        timeout: env.block.time.plus_seconds(PACKET_LIFETIME).into(),
+    };
+    println!("Created IBC message for sending");
+
+    Ok(Response::default()
+        .add_attribute(format!("method"), "execute_send_message")
+        .add_attribute(format!("channel"), channel)
+        .add_attribute("receiving_kernel_address", channel_info.kernel_address)
+        .add_attribute("chain", destination_chain)
+        .add_message(msg))
+}
+
+pub(crate) fn handle_ibc_transfer_funds(
+    deps: DepsMut,
+    info: MessageInfo,
+    env: Env,
+    _ctx: Option<AMPPkt>,
+    message: AMPMsg,
+    channel_info: ChannelInfo,
+) -> Result<Response, ContractError> {
+    let AMPMsg {
+        recipient,
+        message,
+        funds,
+        ..
+    } = message;
+
+    //check if chain is provided
+    let chain = recipient.get_chain().ok_or(ContractError::InvalidPacket {
```

Comment:
We're already doing this in `handle_ibc`, we can add it as a variable to this function